### ISSUE TITLE
Correct library for DS3231

### ISSUE THREAD

**Issue by smz** [smz/Arduino-RTCtime#4]
See: https://github.com/Makuna/Rtc/issues/49

Thanks @mrwgx3

**Comment by smz** [comment]
@mrwgx3, I'm answering here as I don't want to interfere with the the thread regarding the @makuna's library.

Contrary to what @makuna does in his library, here I just have an `GetTemperature()` method that simply returns the floating point temperature.

If I'm not mistaken, to correct my code according to your findings, I should modify my code in `GetTemperature()` from:

```c++
float degrees = (float) _wire.read();
// fraction is just the upper bits
// representing 1/4 of a degree
uint8_t fract = (_wire.read() >> 6) * 25;

degrees += (float) fract / ((degrees < 0) ? -100.0f : 100.0f) ;
return degrees;
```
to:
```c++
uint8_t mstemp = _wire.read();
uint8_t lstemp = _wire.read();
int16_t scaledtemp = mstemp << 8 | lstemp;  // This should be the same as (mstemp << 8) + lstemp
return ((float) scaledtemp  / 256.0f );
```
Is the above correct?

Unfortunately I don't have  a DS3231 at hand right now (_mine died sometime ago_), so I will be really grateful if you could give it a try...

And of course yes, a PR will be appreciated too!

**Comment by smz** [comment]
... and to save a couple of precious bytes and a couple of instructions, it could probably be something like:
```c++
union ds3231temp_t {
  int16_t v;
  struct {
    uint8_t hi;
    uint8_t lo;
  } b;
} ds3231temp;

ds3231temp.b.hi = _wire.read();
ds3231temp.b.lo = _wire.read();

return (float) ds3231temp.v / 256.0f;
```

**Comment by mrwgx3** [comment]
Per your request, here are some test results...
```
float RtcDS3231::GetTemperature ( void )
{
/*
// Code used in my RTC library does not use the Wire library,
// hence I do my own buffering

// Read scaled 16-bit 2's complement integer
// Result returned in bufRtc[0 through 1]
   readDev( RTC_REG_TEMPERATURE, RTC_SIZE_TEMPERATURE );

// Unscaled float temperature
   return(  (int16_t)( (bufRtc[ 0 ] << 8) + bufRtc[ 1 ] ) / 256.0f  );
*/

// Code equivalent for WIRE library
// Read scaled 16-bit 2's complement integer
   Wire.beginTransmission( RTC_I2C_ADDR  );  // For DS3231,  0x68
   Wire.write( RTC_REG_TEMPERATURE );        // ..........., 0x11
   Wire.endTransmission();
   Wire.requestFrom(RTC_I2C_ADDR, 2 );

/* TEST RESULTS

// This works fine
   uint8_t mstemp = Wire.read();
   uint8_t lstemp = Wire.read();
   int16_t scaledtemp = mstemp << 8 | lstemp;  // This should be the same as (mstemp << 8) + lstemp
   return ((float) scaledtemp  / 256.0f );

// --------------------------------

// Doesn't work for the ESP8266, as its byte order is LITTLE ENDIAN
   union ds3231temp_t {
     int16_t v;
     struct {
       uint8_t hi;     // BIG ENDIAN
       uint8_t lo;
     } b;
   } ds3231temp;

   ds3231temp.b.hi = Wire.read();
   ds3231temp.b.lo = Wire.read();
   return (float) ds3231temp.v / 256.0f;

// --------------------------------

// This does work for the ESP8266
   union ds3231temp_t {
     int16_t v;
     struct {
       uint8_t lo;      // LITTLE ENDIAN
       uint8_t hi;
     } b;
   } ds3231temp;

   ds3231temp.b.hi = Wire.read();
   ds3231temp.b.lo = Wire.read();
   return (float) ds3231temp.v / 256.0f;

// --------------------------------

// As for saving bytes, one would thinks this would be close to optimal.
// It does not work, howver, as Wire.read() sequence order is evaluated
// incorrectly
   return(  (int16_t)( (Wire.read() << 8) | (uint8_t)Wire.read() ) / 256.0f  );
*/

// This works because Wire.read() sequence order is enforced
// If you don't need the raw temperature elsewhere, I'd go with this endian independent method
   uint8_t mstemp = Wire.read();
   return(  (int16_t)( (mstemp << 8) | (uint8_t)Wire.read() ) / 256.0f  );

} //GetTemperature
```

**Comment by mrwgx3** [comment]
Have generated pull-request #5

**Comment by smz** [comment]
@mrwgx3: thank you, thank you, thank you!

Gosh... I didn't thought about different processors having a different endianess: thanks for that too!

I'm going to merge your PR and in a few days (after having reviewed if there anything else I can fix/upgrade) I'll release a new version incorporating your fixes.

Best regards,

Sergio

**Comment by smz** [comment]
Closing this issue as we have a valid PR (https://github.com/smz/Arduino-RTCtime/pull/5)

**Comment by smz** [comment]
@mrwgx3 P.S.: I've noticed that for making your PR you have directly modified the "master" branch of your fork. Normally it would be better to create a new branch in your fork, modify it and then issue a PR using the newly created and modified branch...

No problem for me, but I think that having directly modified your master branch you can now have some issue syncing your fork with my origin ...

**Comment by mrwgx3** [comment]
Thanks for the tip, @smz!
Still learning the ropes...
I have a couple other pull-requests which will have the same problem, but they haven't been acted upon yet (Makuna for one). For these, my guess would be to cancel the pull-request, remove the fork, re-fork, then proceed as you suggested. As for your pull request, what would you suggest?

**Comment by smz** [comment]
Hello!

As far as regards your PR here, as I said there was no problem for me: already merged!

To reset your fork to the current state of the origin (this project here), and assuming you're using the GitHub Desktop client I suggest you to try the following:

 - Right click on the project name in the left column and choose "OPen in Git Shell", then, in the shell window issue those commands:
 - `git checkout master`
 - `git remote update`
 - `git reset --hard Arduino-RTCtime/master`
 - `git push origin +master`

If the above works (_it should..._), good! If not, just delete the fork and re-fork again. This would be a problem only if you had any other unmerged branch in your forked repo, bust as this is not the case... no problem!

As far as regards your PR in Makuna's project I've noticed that there is quite a mess... unrelated files (_especially some examples_) **seems** to have been completely changed while thy are not and should not. I don't know the cause, maybe differences in the newline terminators (`LF` vs `CR` or `CRLF`), but I strongly suspect this has something to do with the fact that the PR has not yet been merged...

**Comment by mrwgx3** [comment]
Thanks for your suggestions!
Checking out your observations regarding Makuna pull-request, will be back with commentary shortly

**Comment by mrwgx3** [comment]
Part of the mess comes from my practicing with local repositories on  with 'git'. I did not know that pushing the repository out to github would include all my practice changes. Next time I'll upload intact files, and only once.

As far as the examples go, I created a new example called 'DS3231_Temperature.ino', which exercised all the changes I made to the RtcTemperature class, assuming that Makuna would only include it if he thought it useful. It was derived from 'DS3231_Simple', which had a minor typo that I corrected. I then went and fixed the typo in all the examples which shared it .

Also, I tried to update his documentation wiki by creating a stub in my forked repository, the replacing the stub with a copy of his documentation wiki. This of course breaks any self-referencing links, but this caused no problems on the pages I updated as they had none. I figured he could reference the new wiki and cut/paste the corrections into his, as I don't see any method of merging a wiki repository.

Does this explain the confusion that you see?

Maybe this explains why Makuna hasn't processed my pull-request yet

**Comment by mrwgx3** [comment]
```
Maybe this explains why Makuna hasn't processed my pull-request yet
```
If github doesn't have the ability so select portions of pull-request,  I can see that I may have caused Makuna some real headaches. As I have all my stuff locally, it wouldn't be too difficult to cancel the original pull request, re-fork and change with a much cleaner version. What do you think?

**Comment by smz** [comment]
As a general rule I think it is always safer (_and appreciated_) to keep different functional changes in different PRs (_and hence different branches of your fork_) so that the project owner can decide which PR to merge and which not, so yes I'm afraid you're giving Makuna some headaches with your PR, although he is the only one who cant truly say that...

Anyway I'm afraid: there is something more: if you look at your PR (https://github.com/Makuna/Rtc/pull/50/files) you'll see that:

 - The **entire**  `examples/DS3231_Simple/DS3231_Simple.ino` seems to have changed, every single line of it, not just some line with minor typo, and that's really strange. This is where I suspect different line terminators (or spaces vs `TABs` ??) might be involved...
- You also added the printing of a space in `examples/DS3231_StoreIt/DS3231_StoreIt.ino`. Good or bad I think this should had been a different PR.
- In `src/RtcDS3231.h` you added some newlines in unrelated places (_and the original coder might like it or not_...)

My suggestion is to ask Makuna what he think is best to do with your PR, whether to close it, modify it or submit new separate ones...

**Comment by mrwgx3** [comment]
Thanks for the commentary. I've decided to just redo the pull-request correctly, now that I gotten some good advice. Thanks for you help. Let me know what you think about the 2nd pass.

P.S. Regarding  ```examples/DS3231_Simple/DS3231_Simple.ino```,  I believe I caused this by just simply uploading the file I had locally, hence every line changed.

**Comment by smz** [comment]
No problem: I've been helped too in my first times in GitHub. Learning how to operate here is a quite steep learning curve, but you'll come to really appreciate it, trust me!

> P.S. Regarding examples/DS3231_Simple/DS3231_Simple.ino, I believe I caused this by just simply uploading the file I had locally, hence every line changed.

Yes, that too can explain the fact that it seems to have been completely changed!

Do you mind if I clean-up this thread deleting this last posts which are not related to this project?
Have you copied my instructions for resetting your fork?

**Comment by smz** [comment]
BTW, You are using the Desktop Client, aren't you? If not... you definitely should!!

**Comment by mrwgx3** [comment]
```
Do you mind if I clean-up this thread deleting this last posts which are not related to this project?
Have you copied my instructions for resetting your fork?
```
I have now, and yes, delete the last posts including this one.
Will install the Desktop client and give it a go.
Thanks

P.S.  As for proper introductions, my name is Layne Wright, and I live in Westminster, CO, a suburb of Denver.

**Comment by smz** [comment]
Nice to meet you Layne, I'm Sergio Manzi from Venice, Italy!

If you ever want to get in touch directly, my email is [my-github-id]@[my-github-id].it

**Comment by smz** [comment]
Conversation locked: please open a new issue if you need...
